docs/registry/README.md

```diff
@@ -0,0 +1,193 @@
+# Registry Proposal
+
+## Objective
+
+Design an **initial** version of the Registry that can support discoverability of 
+the different event types that can be consumed from the eventing mesh.
+
+## Requirements
+
+Our design revolves around the following core requirements:
+
+1. We should have a Registry per namespace to enforce isolation.
+2. The Registry should contain the event types that can be consumed from
+the eventing mesh. If an event type is not ready for consumption, we should explicitly indicate so.
+3. The event types stored in the Registry should contain all the required information 
+for a consumer to create a Trigger without resorting to some other OOB mechanism.
+
+## Design Ideas
+
+### EventType CRD
+
+We propose having a namespaced EventType CRD. Here is an example of how a CR would look like:
+
+```yaml
+apiVersion: eventing.knative.dev/v1alpha1
+kind: EventType
+metadata:
+  name: repopush
+spec:
+  type: repo:push
+  source: my-user/my-repo
+  schema: my-schema
+  broker: default
+```
+
+
+- The `name` of the EventType is advisory, non-authoritative. Given that Cloud Event types can 
+contain characters that may not comply with Kubernetes naming conventions, we will (slightly) 
+modify those names to make them K8s-compliant, whenever we need to generate them. 
+
+- `type` is authoritative. This refers to the Cloud Event type as it enters into the eventing mesh. 
+
+- `source`: an identifier of where we receive the event from. This might not necessarily be the Cloud Event source 
+attribute. If we receive the event from our receive adaptors, the info might come in a Cloud Event custom extension (e.g., from).
+
+- `schema` is a URI with the EventType schema. It may be a JSON schema, a protobuf schema, etc. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. 
+
+### EventType Instantiation
+
+We foresee the following ways of populating the Registry:
+
+**1. Event Source CR installation**
+
+Upon installation of an Event Source CR, the source will register its EventTypes.
+
+Example:
+
+```yaml
+apiVersion: sources.eventing.knative.dev/v1alpha1
+kind: GitHubSource
+metadata:
+  name: github-source-sample
+  namespace: default
+spec:
+  eventTypes:
+    - push
+    - pull_request
+  ownerAndRepository: my-other-user/my-other-repo
+  accessToken:
+    secretKeyRef:
+      name: github-secret
+      key: accessToken
+  secretToken:
+    secretKeyRef:
+      name: github-secret
+      key: secretToken
+  sink:
+    apiVersion: eventing.knative.dev/v1alpha1
+    kind: Broker
+    name: default
+
+```
+ 
+By applying this file, two EventTypes will be registered, with types `push` and `pull_request`, 
```

Comment:
The flow here feels a bit off. Maybe I'm doing it wrong :-)  but in my very simple demos my flow is this:
- create my function/sink
- create a github event source to a) make the subscription/webhook in github, and b) connect it up to my sink

So, I sort of view the creation of the github event source as creating a subscription. In order to do this I would need to know what event types to include. I'm assuming the point of the registry is to provide me the list of event types I can choose from. So, populating the registry from the creation of an instance of a github event source feels backwards to me.

Reply:
I guess we are thinking of the problem in the exact opposite way :) and I may be wrong. 
My steps would look like this:

1- I want to receive pull_requests from my **nachocano/eventing** github repo. I then create a GitHub CR for that.

2- This auto-populates the registry, which will tell me what are the *exact* type and source fields I will be able to create triggers on. And also from which *broker* I will be able to consume the event from, and optionally the event schema. Note that the EventType will be added to the registry if your source is pointing to a Broker sink. If you just send stuff to a non-Broker sink, then there won't be an EventType created in the registry. In other words, registry goes along with broker/trigger. 

Anyway, this comes in handy for at least few reasons: 
  * If we are changing the github type from **pull_request** to **dev.knative.source.github.pull_request**, then you will be able to know the type you should filter on. 
  * Also the cloud event source field is somewhat useless. For example, for github pull requests, our current source sends something like: https://github.com/nachocano/eventing/pull/<pull_id>. It is useless to have an exact match on that field. You service don't know that in advance, but I guess with advanced filtering we can account for that. We just care about having the source as **nachocano/eventing**, which is the "source" field that will be added to the registry.
  * You will have a **Ready** status indicating whether that EventType is actually ready for consumption. What if the Broker is broken, or someone mistakenly deleted the source. You will still think that your service is able to consume github pull requests, but it actually doesn't.

I guess this point goes to the fact that you don't need an OOB mechanism to see what you can consume. Let's say another user comes in later on, and she wants to do the same, listen to the same pull_requests from the same repo for some auditing function. She will then create a new source for that? No! she will just look at the registry, see that a source for that already exists, and wire up her function.

3- Then you implement your function.

4- Then you create the Trigger to filter on type: ****dev.knative.source.github.pull_request** and source **nachocano/eventing**, and point it to your function.

Again, I may be wrong, but thanks for comment so that we move the conversation forward.

Reply:
> 1- I want to receive pull_requests from my nachocano/eventing github repo. I then create a GitHub CR for that.

Right here is a big diff/decision point. What event types do you ask for when you create this GithubEventSource? Whatever you pass into this source is what'll be used on the github subscription. If all I care about are "push" events I'd prefer to have Github send just one type of event and not all of them - even if a Broker is in the picture.

Your assumption of a broker being in the system might not be correct. I'm not saying it's not a valid usecase, but there will be times when the eventSource talks directly to the function w/o a broker.

> I guess this point goes to the fact that you don't need an OOB mechanism to see what you can consume.

I'd phrase this differently. This forces me to NOT look at Github for the info. This forces me to look at Kn and have a Kn version of my function. And that concerns me.

> Let's say another user comes in later on, and she wants to do the same, listen to the same pull_requests from the same repo for some auditing function. She will then create a new source for that? No! she will just look at the registry, see that a source for that already exists, and wire up her function.

This goes to something that I wanted to discuss as part of the broader Eventing discussion, and I think it's related to the Registry, what is the user flow? What you describe here is (basically) someone subscribing to a Broker, not Github. I don't view it that way. I think in many cases people will be thinking about subscribing to their event producer, and whether there's a broker in the picture is an impl detail.  Or to put it another way, I think the user flow should be the same regardless of whether this user is first or second w.r.t. wanting events from github. If the system can optimize things and get them to share sources, great, but that's hidden from the user.

Reply:
> Right here is a big diff/decision point. What event types do you ask for when you create this GithubEventSource? Whatever you pass into this source is what'll be used on the github subscription. If all I care about are "push" events I'd prefer to have Github send just one type of event and not all of them - even if a Broker is in the picture.

Yes, we agree on this one. I think I'm not saying something different. You create the github webhook only listening to pushes. That is how is currently being done.

Reply:
> Your assumption of a broker being in the system might not be correct. I'm not saying it's not a valid usecase, but there will be times when the eventSource talks directly to the function w/o a broker.

Agree, the event source can talk directly to a function, and there might not be a broker. In that case, I'm not considering a registry at all (which might be wrong).
What I was trying to say is that if *there is* a broker, then we might be better off by having a registry for that broker.

Reply:
> whether there's a broker in the picture is an impl detail

Hmm, I thought channels were implementation details that we wanted to abstract away with the new Broker model.

Reply:
> > Right here is a big diff/decision point. What event types do you ask for when you create this GithubEventSource? Whatever you pass into this source is what'll be used on the github subscription. If all I care about are "push" events I'd prefer to have Github send just one type of event and not all of them - even if a Broker is in the picture.

> Yes, we agree on this one. I think I'm not saying something different. You create the github webhook only listening to pushes. That is how is currently being done.

I don't think we're on the same page because you start off with:

> 1- I want to receive pull_requests from my nachocano/eventing github repo. I then create a GitHub CR for that.
> 2- This auto-populates the registry, which will tell me what are the exact type and source fields I will be able to create triggers on. 

Which means, I create the Github event source with just "push" events. The registry is populated with just the "push" event.  And then later on you talk about someone creating "just triggers" based on what's in the registry, as if they're subscribing to Kn/brokers, not event producers. Which means if all they look at w.r.t. what they get from Github is what's in the Registry they'll only see "push", not everything that's available from Github. 

Now if when you said "this auto-populates the registry" you meant it populates it with all Github event types, and not just "push" then creating a trigger for anything other than "push" won't work because the subscription to Github is only for push. The creation of a trigger needs to change the github subscription to ask for this new event type, which probably means they're not creating a "trigger" as much as they're really creating a Github event source or some other action outside the notion of just a trigger (which is just the tail end of the flow)

Reply:
> > Your assumption of a broker being in the system might not be correct. I'm not saying it's not a valid usecase, but there will be times when the eventSource talks directly to the function w/o a broker.

> Agree, the event source can talk directly to a function, and there might not be a broker. In that case, I'm not considering a registry at all (which might be wrong).

I really would like to have some real-time chats about the UX here because it's not clear to me if all users setup their triggers/subscriptions the same way or if there's a high-order decision that involves them knowing that a broker is in the picture. And, if there are two different flows, do we expect the user to discover the list of available event types differently?

I guess I see a couple of different flows here:
1 - an enterprise has strict control over what events are seen by Sinks, and in that case the brokers act more like event producers in the sense that they are the source of truth w.r.t. what events can be seen (asking Github doesn't make as much sense here since its a superset of event types that can be seen)
2 - the user examines what already setup in their cluster (e.g. what event sources are there) and wants to just reuse that infrastructure to add more Sinks (triggers) to existing event sources. This is similar to 1, except here the user is choosing to not have a generic install process and is setting up something based off of what they currently see installed
3 - I have this vision where a user tells Kn something like "I want these event types from Github, make it happen", and in that case it'll setup whatever EventSources and Triggers are needed (or reuse existing ones when possible). This allows for the user to do the same type of setup actions whether they're the first one to touch the system or the last.  In some ways I can see us creating a high-level/abstract resource that will auto-generate event sources and triggers based on whatever the user requested, so the user doesn't need to explicitly check what's already deployed.  Or we could just blindly create it all each time, even if that means dups :-)

Reply:
> 3 - I have this vision where a user tells Kn something like "I want these event types from Github, make it happen", and in that case it'll setup whatever EventSources and Triggers are needed (or reuse existing ones when possible). This allows for the user to do the same type of setup actions whether they're the first one to touch the system or the last. In some ways I can see us creating a high-level/abstract resource that will auto-generate event sources and triggers based on whatever the user requested, so the user doesn't need to explicitly check what's already deployed. Or we could just blindly create it all each time, even if that means dups :-)

This idea is pretty cool, and probably something we should try to aim for. Having said that, IMHO I think it might be too ambitious to try to have it done for a first cut. I see 1 and 2 quickly doable (which is basically this PR + the eventing sources one). Get it out there rather soon-ish, and start collecting feedback from users. 

I do agree that users will have to do some more steps and not everything will be automatically created, but it's a start towards potentially the 3rd flow. Sort of what happened with Channels and Broker/Trigger. Users were complaining that they needed a higher level construct to have a better UX, and we then went ahead and added the new model. 

Thanks for your input @duglin , and happy to chat anytime over the week. 
It'd be great to collect some feedback from the other folks as well. I guess they will start commenting over the week.

Reply:
@duglin I agree with your vision, but I think it won't (and should not) happen in the cluster/namespace. In the use case discussion there were several people ( I was one of them 😄 ) mentioning that they already have big message brokers with an existing topic structure for events outside the cluster. We maintain an event catalog for that, and I am a bit worried about replication and consistency issues, when materializing available event types into each namespace that is consuming events. @duglin Maybe we need an Open Event Broker Catalog. 😉 
I also think the term _source_ in knative is misleading, as it is about setting up some entry point to the cluster. So it's a bit like calling an ingress _client_.